cmd/zed/zed.d/statechange-blinkled.sh

```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
+#
+# Turn off/on the VDEV's enclosure fault LEDs when the pool's state changes.
+#
+# Turn LED on if the VDEV becomes faulted/degraded, and turn it back off when
+# it's healthy again.
+#
+# Script does nothing if it's unable to find the LED sysfs entries for the VDEV.
+#
+[ -f "${ZED_ZEDLET_DIR}/zed.rc" ] && . "${ZED_ZEDLET_DIR}/zed.rc"
+. "${ZED_ZEDLET_DIR}/zed-functions.sh"
+
+[ -n "${ZEVENT_VDEV_UPATH}" ] || exit 9
+
+if [ "${ZED_USE_ENCLOSURE_LEDS}" != "1" ] ; then
+	exit 0
+fi
+
+if [ ! -d /sys/class/enclosure ] ; then
+	exit 0
+fi
+
+# Turn on off enclosure LEDs
+function led
+{
+	name=$1
+	val=$2
+
+	# We want to check the current state first, since writing to the
+	# 'fault' entry always always causes a SES command, even if the
+	# current state is already what you want.
+	if [ -e /sys/block/$name/device/enclosure_device*/fault ] ; then
+
+		# We have to do some monkey business to deal with spaces in
+		# enclosure_device names.  I've seen horrible things like this: 
+		#
+		# '/sys/block/sdfw/device/enclosure_device:SLOT 43 41  /fault'
+		#
+		# ...so escape all spaces.
+		file=`ls /sys/block/$name/device/enclosure_device*/fault | sed 's/\s/\\ /g'`
+
+		current=`cat "$file"`
+
+		# On some enclosures if you write 1 to fault, and read it back,
+		# it will return 2.  Treat all non-zero values as 1 for
+		# simplicity.
+		if [ "$current" != "0" ] ; then
+			current=1
+		fi
+
+		if [ "$current" != "$val" ] ; then
+			# Set the value twice.  I've seen enclosures that were
+			# flakey about setting it the first time.
+			echo $val > "$file"
+			echo $val > "$file"
+		fi
+	fi
+}
+
+function process {
+	# path=/dev/sda, fault=[FAULTED|DEGRADED|ONLINE]
+	path=$1
+	fault=$2
+	name=`basename $path`
+
+	if [ -z "$name" ] ; then
+		continue
+	fi
+
+	if [ "$fault" == "FAULTED" ] || [ "$fault" == "DEGRADED" ] ; then
+		led $name 1
+	else
```

Comment:
If UNAVAIL is a choice, maybe this should be restricted to ONLINE (or !UNAVAIL) so that FAULTED->UNAVAIL does not clear the light. Likewise, while OFFLINE should not *light* the light, FAULTED->OFFLINE should not *clear* the light.